### ISSUE TITLE
Spark 1.3 -- Error in retrieving an array element from Elastic Search

### ISSUE THREAD

**Issue by amajumdarinfa** [elastic/elasticsearch-hadoop#529]
Hi
Environment: Spark 1.3.0
 Elasticsearch: 1.4.4

Inside ElasticSearch I have a JSON document with a JSON array. I run a spark application that loads the ElasticSearch Type as a temp table and runs Hive Thrift Server 2. In the client side I have a simple JDBC client that executes a simple SQL statement like "select * from X". The spark application fails to verify the resultset due to the existence of the JSON array and throws the following exception.

```
Exception in thread "main" org.apache.hive.service.cli.HiveSQLException: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 2.0 failed 1 times, most recent failure: Lost task 0.0 in stage 2.0 (TID 10, localhost): java.lang.ClassCastException: scala.collection.convert.Wrappers$JListWrapper cannot be cast to java.lang.String
       at org.elasticsearch.spark.sql.ScalaEsRow.getString(ScalaEsRow.scala:58)
       at org.apache.spark.sql.columnar.StringColumnStats.gatherStats(ColumnStats.scala:208)
       at org.apache.spark.sql.columnar.NullableColumnBuilder$class.appendFrom(NullableColumnBuilder.scala:56)
       at org.apache.spark.sql.columnar.NativeColumnBuilder.org$apache$spark$sql$columnar$compression$CompressibleColumnBuilder$$super$appendFrom(ColumnBuilder.scala:87)
       at org.apache.spark.sql.columnar.compression.CompressibleColumnBuilder$class.appendFrom(CompressibleColumnBuilder.scala:78)
       at org.apache.spark.sql.columnar.NativeColumnBuilder.appendFrom(ColumnBuilder.scala:87)
       at org.apache.spark.sql.columnar.InMemoryRelation$$anonfun$3$$anon$1.next(InMemoryColumnarTableScan.scala:135)
       at org.apache.spark.sql.columnar.InMemoryRelation$$anonfun$3$$anon$1.next(InMemoryColumnarTableScan.scala:111)
       at org.apache.spark.storage.MemoryStore.unrollSafely(MemoryStore.scala:249)
       at org.apache.spark.CacheManager.putInBlockManager(CacheManager.scala:172)
       at org.apache.spark.CacheManager.getOrCompute(CacheManager.scala:79)
       at org.apache.spark.rdd.RDD.iterator(RDD.scala:242)
       at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:35)
       at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:277)
       at org.apache.spark.rdd.RDD.iterator(RDD.scala:244)
       at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:35)
       at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:277)
       at org.apache.spark.rdd.RDD.iterator(RDD.scala:244)
       at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:68)
       at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:41)
       at org.apache.spark.scheduler.Task.run(Task.scala:64)
       at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:203)
       at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
       at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
       at java.lang.Thread.run(Thread.java:745)

Driver stacktrace:
       at org.apache.hive.jdbc.Utils.verifySuccess(Utils.java:231)
       at org.apache.hive.jdbc.Utils.verifySuccessWithInfo(Utils.java:217)
       at org.apache.hive.jdbc.HiveStatement.execute(HiveStatement.java:254)
       at com.infa.sats.daoTest.Test.main(Test.java:34)
```

The JSON document looks like this -->
```json
{
    "username": "CCC",
    "firstName" : "CC",
    "lastName" : "C",
    "fullName" : "CCC",
    "email" : "CCC2@infa.com",
    "groups" : ["Group_111"] 
}
```
Note that if I do not have "groups" as an array the issue goes away. 

Is it a known issue? Do you suggest any work-around?

Thanks
Abhijit

**Comment by costin** [comment]
You're not mentioning the es-hadoop version used.

Likely you are running into this issue with Spark SQL since it requires types to be static. In ES a field can have 1 or multiple values however since the schema cannot infer this, it results in exceptions like this.

**Comment by amajumdarinfa** [comment]
Thanks Costin!

THe es-hadoop version is 2.1.0.

Ok, that means the ES Type must have a schema in order to successfully integrate an array field with Spark SQL.

**Comment by costin** [comment]
No, rather that ES is more flexible than Spark SQL allows. The solution
here is to add an extra option in SparkSQL to map certain fields as
arrays/list instead of simple values.

**Comment by costin** [comment]
Fixed through #482  #484 . Thanks.
